File: 199_Binary Tree Right Side View.java
// 时间复杂度：O(n)
// 空间复杂度：O(n)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
// Method 1: DFS
class Solution {
    private void dfs(TreeNode node, int depth, List<Integer> res) {
        if (node == null) return;
        if (res.size() == depth) res.add(node.val);
        dfs(node.right, depth+1, res);
        dfs(node.left, depth+1, res);
    }

    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        dfs(root, 0, res);
        return res;
    }
}

// Method 2: BFS
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        Queue<TreeNode> q = new LinkedList<>();

        if (root != null) q.offer(root);

        while (!q.isEmpty()) {
            int length = q.size();
            for (int i = 0; i < length; i++) {
                TreeNode cur = q.poll();
                if (i == length - 1) res.add(cur.val);
                if (cur.left != null) q.offer(cur.left);
                if (cur.right != null) q.offer(cur.right);
            }
        }

        return res;
    }
}

File: 199_Binary Tree Right Side View.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        # BFS
        q = deque()
        res = []

        if root: q.append(root)

        while q:
            length = len(q)

            for i in range(length):
                node = q.popleft()
                if i == length - 1: res.append(node.val)
                if node.left: q.append(node.left)
                if node.right: q.append(node.right)

        return res